presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java

```diff
@@ -75,21 +75,50 @@ public void testJoin()
     }
 
     @Test
-    public void testSetOperations()
+    public void testUnion()
     {
         statisticsAssertion.check("SELECT * FROM nation UNION SELECT * FROM nation",
-                // real count is 25, estimation cannot know all rows are duplicate. TODO create better test for UNION
-                checks -> checks.estimate(OUTPUT_ROW_COUNT, relativeError(0., 1.)));
+                // real count is 25, estimation cannot know all rows are duplicate.
+                checks -> checks.estimate(OUTPUT_ROW_COUNT, relativeError(1, 1)));
 
         statisticsAssertion.check("SELECT * FROM nation UNION ALL SELECT * FROM nation",
                 checks -> checks.estimate(OUTPUT_ROW_COUNT, noError()));
 
+        statisticsAssertion.check("SELECT * FROM orders WHERE o_custkey < 755 OR o_orderstatus = '0' UNION SELECT * FROM orders WHERE o_custkey > 755 OR o_orderstatus = 'F'",
+                checks -> checks.estimate(OUTPUT_ROW_COUNT, relativeError(.3, .35)));
+
+        statisticsAssertion.check("SELECT * FROM orders WHERE o_custkey < 755 OR o_orderstatus = '0' UNION ALL SELECT * FROM orders WHERE o_custkey > 755 OR o_orderstatus = 'F'",
+                checks -> checks.estimate(OUTPUT_ROW_COUNT, defaultTolerance()));
+
+        statisticsAssertion.check("SELECT * FROM orders WHERE o_custkey < 900 UNION SELECT * FROM orders WHERE o_custkey > 600",
+                checks -> checks.estimate(OUTPUT_ROW_COUNT, relativeError(.15, .25)));
+
+        statisticsAssertion.check("SELECT * FROM orders WHERE o_custkey < 900 UNION ALL SELECT * FROM orders WHERE o_custkey > 600",
+                checks -> checks.estimate(OUTPUT_ROW_COUNT, defaultTolerance()));
+    }
+
+    @Test
+    public void testIntersect()
+    {
         statisticsAssertion.check("SELECT * FROM nation INTERSECT SELECT * FROM nation",
-                checks -> checks.estimate(OUTPUT_ROW_COUNT, relativeError(0, 1.)));
+                // real count is 25, estimation cannot know all rows are duplicate.
+                checks -> checks.estimate(OUTPUT_ROW_COUNT, relativeError(.7, .9)));
+
+        statisticsAssertion.check("SELECT * FROM orders WHERE o_custkey < 900 INTERSECT SELECT * FROM orders WHERE o_custkey > 600",
+                // TODO fix INTERSECT stats calculation as custkey values distirbution is pretty linear
```

Comment:
@kokosing could you?

Reply:
I wish I could remember this...
I suspect it is because INTERSECT is replaced with an aggregation and then once it happens it is no longer trivial to estimate stats.